icu4c/source/i18n/smpdtfmt.cpp

```diff
@@ -1480,6 +1481,9 @@ SimpleDateFormat::subFormat(UnicodeString &appendTo,
                             Calendar& cal,
                             UErrorCode& status) const
 {
+    static const int32_t maxIntCount = 10;
+    static const UnicodeString hebr("hebr", 4, US_INV);
```

Comment:
Making this declaration `static const` is an improvement, but I wonder if we can get rid of it altogether.  I only see it used in one spot, where we have ` if (fDateOverride.compare(hebr)==0 && value>HEBREW_CAL_CUR_MILLENIUM_START_YEAR ...)`, and I think you could just change this to `if (fDateOverride == u"hebr" && ...)` and avoid creating a `UnicodeString` at all.  (In addition, using `compare()` instead of `operator==()` to test for equality seems questionable.)

Reply:
This doesn't seem to work:

```c++
smpdtfmt.cpp:1565:27: warning: result of comparison against a string literal is unspecified (use an explicit string comparison function instead) [-Wstring-compare]
        if (fDateOverride == u"hebr" && value>HEBREW_CAL_CUR_MILLENIUM_START_YEAR && value<HEBREW_CAL_CUR_MILLENIUM_END_YEAR) {
                          ^  ~~~~~~~
smpdtfmt.cpp:1565:27: error: invalid operands to binary expression ('const UnicodeString' and 'const char16_t[5]')
        if (fDateOverride == u"hebr" && value>HEBREW_CAL_CUR_MILLENIUM_START_YEAR && value<HEBREW_CAL_CUR_MILLENIUM_END_YEAR) {
            ~~~~~~~~~~~~~ ^  ~~~~~~~
../common/unicode/stringpiece.h:326:1: note: candidate function not viable: no known conversion from 'const UnicodeString' to 'const StringPiece' for 1st argument
operator==(const StringPiece& x, const StringPiece& y);
^
../common/unicode/unistr.h:3978:16: note: candidate function not viable: no known conversion from 'const char16_t[5]' to 'const UnicodeString' for 1st argument
UnicodeString::operator== (const UnicodeString& text) const
```

Reply:
I guess that means you have to do `fDateOverride == UnicodeString(u"hebr")`, which is kind of ugly.  You could do that or stick with the static declaration.  Since it's only used in the one spot, I'm inclined to say you should put the declaration inline, but it's up to you.  Regardless, I'd do `UnicodeString(u"hebr")` instead of `UnicodeString("hebr", 4, US_INV)`.  The second version is old and does a conversion-- it dates to long before Unicode string literals were supported in C++ (or in the version of C++ ICU was using, which was always way behind the current one).  I'm pretty sure the constructor that takes a `const UChar*` just wraps the `UnicodeString` object around that storage, with no copying or conversion, and should be fairly fast.

Reply:
Looking at the code it does seem to call `doAppend`:

```cpp
UnicodeString::UnicodeString(const char16_t *text) {
  fUnion.fFields.fLengthAndFlags = kShortString;
  doAppend(text, 0, -1);
}
```

I'll switch to the new style but keep it as `static` to prevent the construction overhead.